src/anyp/Uri.cc

```diff
@@ -81,6 +103,50 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
+
+    Parser::Tokenizer tok(buf);
+    SBuf token;
+    static const auto unencodedOctets = CharacterSet("unencoded", "%").complement();
+
+    tok.prefix(token, unencodedOctets);
+    output.append(token);
+
+    while (!tok.atEnd())
+    {
```

Comment:
After "refactor step 1a and 1b" we now have `unencoded` characters handled first and at each loop ending.

Refactor step 2: This code can be de-duplicated by moving the unencoded string handling into the start of the loop (and up from the back).

```suggestion
    {
        static const auto unencodedOctets = CharacterSet("unencoded", "%").complement();
        output.append(tok.prefix("unencoded", unencodedOctets));

```

Reply:
Adopted.